analyses/long-format-table-utils/annotator/download-annotation-data.R

```diff
@@ -0,0 +1,205 @@
+# from.
+#
+# This only works if the working directory is OpenPedCan-analysis or a
+# subdirectory of OpenPedCan-analysis
+#
+# root_dir is the absolute path of OpenPedCan-analysis
+tryCatch(
+  {
+    root_dir <- rprojroot::find_root(rprojroot::has_file(".git/index"))
+  },
+  error = function(err_cond) {
+    # adapted from http://adv-r.had.co.nz/Exceptions-Debugging.html
+    err_cond$message <- paste0(
+      err_cond$message,
+      "\nTry re-running this script with working directory as ",
+      "OpenPedCan-analysis or a subdirectory of OpenPedCan-analysis.\n")
+    stop(err_cond)
+  }
+)
+
+input_data_dir <- file.path(root_dir, "data")
+
+output_data_dir <- file.path(
+  root_dir, "analyses", "long-format-table-utils", "annotator",
+  "annotation-data")
+
+if (!dir.exists(output_data_dir)) {
+  dir.create(output_data_dir)
+}
+
+
+
+# Read input data --------------------------------------------------------------
+# ensg hugo rmtl mappings
+ensg_hugo_rmtl_df <- dplyr::distinct(
+  readr::read_tsv(file.path(input_data_dir, "ensg-hugo-rmtl-v1-mapping.tsv"),
+                  col_types = readr::cols(.default = readr::col_guess())))
+
+# assert all ensg_ids and gene_symbols are not NA
+if (!identical(sum(is.na(ensg_hugo_rmtl_df$ensg_id)), as.integer(0))) {
+  stop(paste0("Found NA in ensg-hugo-rmtl-v1-mapping.tsv ensg_id.\n",
+              "Check if PedOT release data are downloaded properly.\n",
+              "If data is downloaded properly, submit a GitHub data question."))
+}
+
+if (!identical(sum(is.na(ensg_hugo_rmtl_df$gene_symbol)), as.integer(0))) {
+  stop(paste0("Found NA in ensg-hugo-rmtl-v1-mapping.tsv gene_symbol.\n",
+              "Check if PedOT release data are downloaded properly.\n",
+              "If data is downloaded properly, submit a GitHub data question."))
+}
+
+# assert all ensg_id are unique
+if (!identical(length(unique(ensg_hugo_rmtl_df$ensg_id)),
+               nrow(ensg_hugo_rmtl_df))) {
+  stop(paste0("Found duplicated ensg_id in ensg-hugo-rmtl-v1-mapping.tsv.\n",
+              "Check if PedOT release data are downloaded properly.\n",
+              "If data is downloaded properly, submit a GitHub data question."))
+}
+
+# Download data from https://mygene.info/ --------------------------------------
+message("Retrieve Gene_full_name and Protein_RefSeq_ID from mygene.info...")
+ens_gids <- ensg_hugo_rmtl_df$ensg_id
+
+mg_qres_list <- mygene::queryMany(
+  ens_gids, scopes = "ensembl.gene", fields = c("refseq", "name"),
+  species = "human", returnall = TRUE, return.as = "DataFrame")
+
+found_mg_qres_df <- tibble::as_tibble(
+  mg_qres_list$response[, c("query", "notfound", "name", "refseq.protein")]) %>%
+  tidyr::replace_na(list(notfound = FALSE)) %>%
+  dplyr::filter(!notfound)
+
+# remove rows that have both NA name and NULL refseq.protein
+rm_bnn_found_mg_qres_df <- found_mg_qres_df %>%
+  dplyr::filter(!(is.na(name) & purrr::map_lgl(refseq.protein, is.null)))
```

Comment:
nice - I noticed that there are no genes missing both here which led me to ask whether we have any in `ensg-hugo-rmtl-v1-mapping.tsv` which are missing from this annotation and it looks like there are 17, 110 genes missing:

```
> length(setdiff(ensg_hugo_rmtl_df$ensg_id, rm_bnn_found_mg_qres_df$query))
[1] 17110
```

Is this due to the ENSG or GENCODE version differences we used for RNA-Seq ([v27](https://alexslemonade.github.io/OpenPBTA-manuscript/#gene-expression)) vs what mygene is using? What version do they use? We should try to match or if not flexible, at minimum, add this to the README.

Reply:
The ENSG IDs that have no gene name or protein RefSeq IDs are probably not included, or not stable, or retired in HGNC (source of full name / "Approved name"), or NCBI Gene (source of protein RefSeq IDs), or Ensembl (source of ENSG ID mapping). For example, [ENSG00000106540](https://www.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000106540;r=7:39609717-39610280;t=ENST00000446267) is not available in [HGNC](https://www.genenames.org/tools/search/#!/?query=ENSG00000106540) or [NCBI Gene](https://www.ncbi.nlm.nih.gov/gene/?term=ENSG00000106540). Another example, [ENSG00000069712](https://www.ensembl.org/Homo_sapiens/Gene/Idhistory?g=ENSG00000069712) is retired in Ensembl and has "Placeholder symbol" rather than "Stable symbol" in [HGNC](https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:29192), and it is not available in [NCBI Gene](https://www.ncbi.nlm.nih.gov/gene/?term=ENSG00000069712).  I only picked several cases for checking. If we need to check all genes, I could create a ticket for that purpose, probably by using HGNC and Ensembl release files.

The mygene.info uses the latest Ensembl release and NCBI snapshot, which is checked on a weekly basis, according to <https://docs.mygene.info/en/latest/doc/data.html#data-sources>. The current version is Ensembl release 104 and NCBI snapshot 20210625. More information about source data versions is in JSON format at <https://mygene.info/v3/metadata>.

I think mygene.info does not allow users to specify annotation versions, as there is no version related fields in there API <https://docs.mygene.info/en/latest/doc/query_service.html#available-fields>.

We could match versions with the biomart package, but there are more missing gene full names and protein RefSeq IDs. If we use biomart GENCODE version 27 (Ensembl 90), 38,738 ENSG IDs have gene names, and 19,153 ENSG IDs have protein RefSeq IDs. If we use mygene, 41,192 ENSG IDs have gene names, and 19,222 ENSG IDs have protein RefSeq IDs.

I will add these information in the README.md and note the date of the file being downloaded.

Reply:
To note the biomart code here:

```R
bm_ensembl <- biomaRt::useEnsembl(
  biomart = "ensembl", dataset = "hsapiens_gene_ensembl", version = 90)
# bm_attributes <- biomaRt::listAttributes(bm_ensembl)

# If queried with three attrs, certain ENSG IDs that have descriptions will be
# omitted.
#
# From getBM focumentation, it is not guaranteed to have all columns as
# characters. Convert factors to characters.
bm_ensg_desc_df <- as_tibble(
  biomaRt::getBM(
    attributes = c("ensembl_gene_id", "description"),
    filters = c("ensembl_gene_id"),
    values = ens_gids,
    mart = bm_ensembl)) %>%
  mutate_if(is.factor, as.character) %>%
  filter(!ensembl_gene_id == "", !description == "") %>%
  filter(!is.na(ensembl_gene_id), !is.na(description)) %>%
  rename(Gene = ensembl_gene_id, Gene_full_name = description)

bm_ensg_np_df <- as_tibble(biomaRt::getBM(
  attributes = c("ensembl_gene_id", "refseq_peptide"),
  filters = c("ensembl_gene_id"),
  values = ens_gids,
  mart = bm_ensembl)) %>%
  mutate_if(is.factor, as.character) %>%
  filter(!ensembl_gene_id == "", !refseq_peptide == "") %>%
  filter(!is.na(ensembl_gene_id), !is.na(refseq_peptide))
```

Reply:
OK great, thanks for checking and adding the details to the README! Curious and not in the scope of this PR, if we see appreciable gene expression in any genes which are retired, which may give us reason to re-align using a later ENS version.

Reply:
Thank you for pointing out this issue and clarifying the context!

The notes are added in this commit <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/55/commits/a18347018d0f1af57a42210f221de9a36eec5f4e>, which are rendered by GitHub at <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/blob/a18347018d0f1af57a42210f221de9a36eec5f4e/analyses/long-format-table-utils/README.md>.